### PR TITLE
Make diff a visitor

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by certik** [comment]
Nice, thanks. I need to study this PR. We need to carefully benchmark it, especially on expressions where no caching helps.

File: symengine/derivative.cpp [hunk 1]
```diff
@@ -627,81 +662,98 @@ class DiffImplementation
...
+    }
+
+    void apply(const RCP<const Basic> &b)
+    {
+        auto it = visited.find(b);
+        if (it == visited.end()) {
+            b->accept(*this);
+            insert(visited, b, result_);
+        } else {
+            result_ = it->second;
```
Comment (certik) [inline symengine/derivative.cpp]: This might be very slow for some expressions. We need to benchmark it.

Comment (certik) [reply symengine/derivative.cpp]: Whatever is easier for you. I just don't want to merge anything into master that slows things down.

Comment (isuruf) [reply symengine/derivative.cpp]: I reverted the `don't revisit` part of the PR.

**Comment by certik** [comment]
I ran the S3 and S3a benchmarks in (./benchmarks/symbench) that contain some `diff`, and I can't tell a difference between master and this PR (which is good: no significant slowdown).

Do you have some benchmark with visible speedups?

**Comment by isuruf** [comment]
@certik, did you try 297a403 or 1049f2c ?

**Comment by certik** [comment]
I tried 297a403fd51b1d366cdedde348799a7a69271bda.

**Comment by isuruf** [comment]
297a403 isn't supposed to give a speedup nor a slowdown. If you agree that there's no slowdown, we can merge this PR and revert 297a403 in another PR.

**Comment by certik** [comment]
I was worried that 297a403fd51b1d366cdedde348799a7a69271bda will give slowdowns due to the fact that visitor pattern (two virtual function calls) is more expensive than the previous implementation which only needed one virtual function call.

**Comment by isuruf** [comment]
Sure, but it doesn't seem to give a slowdown in my testing and yours according to https://github.com/symengine/symengine/pull/1583#issuecomment-518407952

**Comment by certik** [comment]
Indeed it doesn't seem to slow things down, so if there is a difference, it's a very minor one.

The advantage of the visitor pattern is that we can now structure all algorithms in SymEngine using it, which simplifies the core. I think `diff` was the last algorithm that didn't use it.

So I am fine merging it as is.

**Pull request merged**